src/node/acme_client.h

```diff
@@ -0,0 +1,1198 @@
+    };
+
+    std::list<Order> active_orders;
+
+    static http::URL with_default_port(
+      const std::string& url, const std::string& default_port = "443")
+    {
+      http::URL r = http::parse_url_full(url);
+      if (r.port.empty())
+      {
+        r.port = default_port;
+      }
+      return r;
+    }
+
+    static std::vector<uint8_t> s2v(const std::string& s)
+    {
+      return std::vector<uint8_t>(s.data(), s.data() + s.size());
+    }
+
+    static std::vector<uint8_t> json_to_bytes(const nlohmann::json& j)
+    {
+      return s2v(j.dump());
+    }
+
+    static std::string json_to_b64url(
+      const nlohmann::json& j, bool with_padding = true)
+    {
+      return crypto::b64url_from_raw(json_to_bytes(j), with_padding);
+    }
+
+    static void convert_signature_to_ieee_p1363(std::vector<uint8_t>& sig)
+    {
+      // Convert signature from ASN.1 format to IEEE P1363
+      const unsigned char* pp = sig.data();
+      ECDSA_SIG* sig_r_s = d2i_ECDSA_SIG(NULL, &pp, sig.size());
+      const BIGNUM* r = ECDSA_SIG_get0_r(sig_r_s);
+      const BIGNUM* s = ECDSA_SIG_get0_s(sig_r_s);
+      int r_n = BN_num_bytes(r);
+      int s_n = BN_num_bytes(s);
+      assert(r_n <= 48 && s_n <= 48);
+      sig = std::vector<uint8_t>(96, 0);
+      BN_bn2bin(r, sig.data() + 48 - r_n);
+      BN_bn2bin(s, sig.data() + 96 - s_n);
+    }
+
+    class JWS : public nlohmann::json::object_t
+    {
+    public:
+      JWS(
+        const nlohmann::json& header_,
+        bool header_is_protected_,
+        const nlohmann::json& payload_,
+        crypto::KeyPair& signer_,
+        bool empty_payload = false) :
+        header_is_protected(header_is_protected_)
+      {
+        LOG_TRACE_FMT("JWS header: {}", header_.dump());
+        LOG_TRACE_FMT("JWS payload: {}", payload_.dump());
+        auto header_b64 = json_to_b64url(header_, false);
+        auto payload_b64 = empty_payload ? "" : json_to_b64url(payload_, false);
+        set(header_b64, payload_b64, signer_);
+      }
+
+      JWS(
+        const nlohmann::json& header_,
+        bool header_is_protected_,
+        crypto::KeyPair& signer_) :
+        JWS(
+          header_,
+          header_is_protected_,
+          nlohmann::json::object_t(),
+          signer_,
+          true)
+      {}
+
+      virtual ~JWS() {}
+
+    protected:
+      bool header_is_protected = true;
+
+      void set(
+        const std::string& header_b64,
+        const std::string& payload_b64,
+        crypto::KeyPair& signer)
+      {
+        auto msg = header_b64 + "." + payload_b64;
+        auto sig = signer.sign(s2v(msg));
+        convert_signature_to_ieee_p1363(sig);
+        auto sig_b64 = crypto::b64url_from_raw(sig);
+
+        (*this)[header_is_protected ? "protected" : "header"] = header_b64;
+        (*this)["payload"] = payload_b64;
+        (*this)["signature"] = sig_b64;
+      }
+    };
+
+    class JWK : public nlohmann::json::object_t
```

Comment:
Why not model this and JWS like the other strongly-typed JSON classes?

Reply:
Not sure what you mean by strongly-typed, but the abstraction just makes sense.

Reply:
(That's not what I would call strongly typed, but that's beside the point.) Yes, I'm sure that would work, but it's an additional copy of the data structure, because you first have to create the C++ struct and then create the json object. In the JWK/JWS case, I will never ever need the C++ struct, only the json.

Reply:
I'll delegate to the others for this, maybe I'm missing something. @eddyashton @jumaffre

Reply:
I think it's preferable to stick to what we already do for all serialisable structures and use the `DECLARE_JSON_...()` macros for this new type too?

Reply:
I don't need a serializable structure here, I just need the json, but if you think it helps I can add one anyways.

Reply:
Ah! I think both of you missed the fact that `class JWS` and `class JWK` have no fields, i.e. nothing to serialize. They are quite literally just json-constructors. If it feels more natural to you, I can also put them into functions that return `json` for instance.

Reply:
That also seems fine to me, and so does a possible free function. If the other facilities offered by the macro (deserialisation and schema generation) aren't needed, I don't see why they must be used.